README
=============
indicator-usb
=============

Cool flash drive unmounter for Ubuntu

For more information, please visit our website:
https://github.com/atareao/indicator-usb
or
http://www.atareao.es

============
Requirements
============

Required dependencies:

    python3,
    python3-gi,
    gir1.2-gtk-3.0,
    gir1.2-appindicator3-0.1,
    gir1.2-notify-0.7

========
Download
========

Download the package:

    $ git clone https://github.com/atareao/indicator-usb.git


debian/changelog
indicator-usb (0.0.5-0extras16.04.3) xenial; urgency=medium

  * First version

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 16 Jun 2016 21:56:45 +0200


src/comun.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of pdf-tools
#
# Copyright (C) 2012-2016 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import locale
import gettext
import sys


def is_package():
    return __file__.find('src') < 0

PARAMS = {'first-time': True,
          'version': '',
          'show-hdd': False,
          'show-net': False
          }

APP = 'indicator-usb'
APPNAME = 'Indicator USB'
APP_CONF = APP + '.conf'
CONFIG_DIR = os.path.join(os.path.expanduser('~'), '.config')
CONFIG_APP_DIR = os.path.join(CONFIG_DIR, APP)
CONFIG_FILE = os.path.join(CONFIG_APP_DIR, APP_CONF)


if is_package():
    ROOTDIR = '/opt/extras.ubuntu.com/indicator-usb/share/'
    LANGDIR = os.path.join(ROOTDIR, 'locale-langpack')
    APPDIR = os.path.join(ROOTDIR, APP)
    ICONDIR = os.path.join(ROOTDIR, 'icons')
    CHANGELOG = os.path.join(APPDIR, 'changelog')
    AUTOSTART = os.path.join(APPDIR, 'indicator-usb-autostart.desktop')
else:
    ROOTDIR = os.path.dirname(__file__)
    LANGDIR = os.path.normpath(os.path.join(ROOTDIR, '../po'))
    APPDIR = ROOTDIR
    DATADIR = os.path.normpath(os.path.join(ROOTDIR, '../data'))
    ICONDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/icons'))
    DEBIANDIR = os.path.normpath(os.path.join(ROOTDIR, '../debian'))
    CHANGELOG = os.path.join(DEBIANDIR, 'changelog')
    AUTOSTART = os.path.normpath(
        os.path.join(ROOTDIR, '../data/indicator-usb-autostart.desktop'))
IMAGE = os.path.join(ICONDIR, 'indicator-usb.png')
ICON = os.path.join(ICONDIR, 'indicator-usb.svg')

f = open(CHANGELOG, 'r')
line = f.readline()
f.close()
pos = line.find('(')
posf = line.find(')', pos)
VERSION = line[pos+1:posf].strip()
if not is_package():
    VERSION = VERSION + '-src'
try:
    current_locale, encoding = locale.getdefaultlocale()
    language = gettext.translation(APP, LANGDIR, [current_locale])
    language.install()
    if sys.version_info[0] == 3:
        _ = language.gettext
    else:
        _ = language.ugettext
except Exception as e:
    print(e)
    _ = str
APPNAME = _(APPNAME)


src/configurator.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of indicator-usb
#
# Copyright (C) 2016 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import codecs
import os
import json
import comun


class Configuration(object):
    def __init__(self):
        if not os.path.exists(comun.CONFIG_APP_DIR):
            os.makedirs(comun.CONFIG_APP_DIR)
        elif os.path.isfile(comun.CONFIG_APP_DIR):
            os.remove(comun.CONFIG_APP_DIR)
            os.makedirs(comun.CONFIG_APP_DIR)
        self.params = comun.PARAMS
        self.read()

    def get(self, key):
        try:
            return self.params[key]
        except KeyError as e:
            print(e)
            self.params[key] = comun.PARAMS[key]
            return self.params[key]

    def set(self, key, value):
        self.params[key] = value

    def reset(self):
        if os.path.exists(comun.CONFIG_FILE):
            os.remove(comun.CONFIG_FILE)
        self.params = comun.PARAMS
        self.save()

    def set_defaults(self):
        self.params = comun.PARAMS
        self.save()

    def read(self):
        try:
            f = codecs.open(comun.CONFIG_FILE, 'r', 'utf-8')
        except IOError as e:
            print(e)
            self.save()
            f = codecs.open(comun.CONFIG_FILE, 'r', 'utf-8')
        try:
            self.params = json.loads(f.read())
        except ValueError as e:
            print(e)
            self.save()
        f.close()

    def save(self):
        if not os.path.exists(comun.CONFIG_APP_DIR):
            os.makedirs(comun.CONFIG_APP_DIR)
        f = codecs.open(comun.CONFIG_FILE, 'w', 'utf-8')
        f.write(json.dumps(self.params))
        f.close()


src/preferences_dialog.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of indicator-usb
#
# Copyright (C) 2016 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import gi
try:
    gi.require_version('Gtk', '3.0')
except Exception as e:
    print(e)
    exit(1)
from gi.repository import Gtk
import comun
import os
import shutil
import comun
from comun import _
from configurator import Configuration


class PreferencesDialog(Gtk.Dialog):
    def __init__(self, parent):
        #
        Gtk.Dialog.__init__(self,
                            'Indicator-USB | '+_('Preferences'),
                            parent,
                            Gtk.DialogFlags.MODAL |
                            Gtk.DialogFlags.DESTROY_WITH_PARENT,
                            (Gtk.STOCK_CANCEL, Gtk.ResponseType.REJECT,
                                Gtk.STOCK_OK, Gtk.ResponseType.ACCEPT))
        self.set_position(Gtk.WindowPosition.CENTER_ALWAYS)
        self.set_icon_from_file(comun.ICON)
        #
        vbox0 = Gtk.VBox(spacing=5)
        vbox0.set_border_width(5)
        self.get_content_area().add(vbox0)
        # ***************************************************************
        notebook = Gtk.Notebook.new()
        vbox0.add(notebook)
        # ***************************************************************
        vbox11 = Gtk.VBox(spacing=5)
        vbox11.set_border_width(5)
        notebook.append_page(vbox11, Gtk.Label.new(_('General')))
        frame11 = Gtk.Frame()
        vbox11.pack_start(frame11, False, True, 1)
        table11 = Gtk.Table(2, 3, False)
        frame11.add(table11)
        # ***************************************************************
        label11 = Gtk.Label(_('Show disks')+':')
        label11.set_alignment(0, 0.5)
        table11.attach(label11, 0, 1, 0, 1, xpadding=5, ypadding=5)
        self.autostart = Gtk.Switch()
        table11.attach(self.autostart, 1, 2, 0, 1,
                       xpadding=5,
                       ypadding=5,
                       xoptions=Gtk.AttachOptions.SHRINK)
        label12 = Gtk.Label(_('Show disks')+':')
        label12.set_alignment(0, 0.5)
        table11.attach(label12, 0, 1, 1, 2, xpadding=5, ypadding=5)
        self.show_disks = Gtk.Switch()
        table11.attach(self.show_disks, 1, 2, 1, 2,
                       xpadding=5,
                       ypadding=5,
                       xoptions=Gtk.AttachOptions.SHRINK)
        label14 = Gtk.Label(_('Show network folders')+':')
        label14.set_alignment(0, 0.5)
        table11.attach(label14, 0, 1, 2, 3, xpadding=5, ypadding=5)
        self.show_net = Gtk.Switch()
        table11.attach(self.show_net, 1, 2, 2, 3,
                       xpadding=5,
                       ypadding=5,
                       xoptions=Gtk.AttachOptions.SHRINK)
        self.load_preferences()
        self.show_all()

    def load_preferences(self):
        configuration = Configuration()
        self.show_disks.set_active(configuration.get('show-hdd'))
        self.show_net.set_active(configuration.get('show-net'))
        filestart = os.path.join(
            os.getenv("HOME"),
            ".config/autostart/indicator-usb-autostart.desktop")
        self.autostart.set_active(os.path.exists(filestart))

    def save_preferences(self):
        configuration = Configuration()
        configuration.set('show-hdd', self.show_disks.get_active())
        configuration.set('show-net', self.show_net.get_active())
        configuration.save()
        filestart = os.path.join(
            os.getenv("HOME"),
            ".config/autostart/indicator-usb-autostart.desktop")
        if self.autostart.get_active():
            if not os.path.exists(os.path.dirname(filestart)):
                os.makedirs(os.path.dirname(filestart))
            shutil.copyfile(comun.AUTOSTART, filestart)
        else:
            if os.path.exists(filestart):
                os.remove(filestart)

if __name__ == "__main__":
    cm = PreferencesDialog(None)
    if cm.run() == Gtk.ResponseType.ACCEPT:
        print(1)
    cm.hide()
    cm.destroy()
    exit(0)


src/indicatorusb.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of indicator-usb
#
# Copyright (C) 2016 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('AppIndicator3', '0.1')
    gi.require_version('Notify', '0.7')
except Exception as e:
    print(e)
    exit(1)
from gi.repository import Gtk
from gi.repository import GdkPixbuf
from gi.repository import Gio
from gi.repository import AppIndicator3 as AppIndicator
from gi.repository import Notify
import os
import dbus
from configurator import Configuration
from preferences_dialog import PreferencesDialog
import comun
from comun import _

# YUCK: Gtk.image_new_from_gicon does not work, and pynotify does
# not support gicon, so just let the theme choose the icon


def _get_icon_name_from_gicon(gicon):
    assert type(gicon) == Gio.ThemedIcon
    name = "image-missing"
    theme = Gtk.IconTheme.get_default()
    for n in gicon.get_names():
        if theme.lookup_icon(n, Gtk.IconSize.MENU, 0):
            name = n
            break
    return name


class Main:
    def __init__(self):
        if dbus.SessionBus().request_name('es.atareao.indicator-usb') !=\
                dbus.bus.REQUEST_NAME_REPLY_PRIMARY_OWNER:
            print("application already running")
            exit(0)
        self.drives = []
        self.menu = None
        self.load_preferences()
        self.notifications = Notify.Notification.new('', '', None)
        self.indicator = AppIndicator.Indicator.new(
            _('Indicator-USB'),
            _('Indicator-USB'),
            AppIndicator.IndicatorCategory.HARDWARE)
        self.indicator.set_status(AppIndicator.IndicatorStatus.ACTIVE)
        self.indicator.set_icon('media-removable-symbolic')
        self.monitor = Gio.VolumeMonitor.get()
        self.monitor.connect('mount-added', self._add_drive)
        self.monitor.connect('mount-removed', self._del_drive)
        self.refresh()

    def refresh(self):
        self.drives = []
        for m in self.monitor.get_mounts():
            self._add_drive(None, m)

    def _add_drive(self, s, m):
        if m.can_unmount():
            root = m.get_root()
            if (root.get_uri_scheme() == "file" and self.show_hdd) \
               or (not root.is_native() and self.show_net):
                self.drives.append(m)
        self.update()

    def _del_drive(self, s, d):
        self.drives.remove(d)
        self.update()

    def _eject_cb(self, m, result, t):
        # XXX: pynotify does not support gicon
        n = Notify.Notification.new(_('Device can be removed now'),
                                    m.get_name(),
                                    _get_icon_name_from_gicon(m.get_icon()))
        n.show()

    def eject(self, s, m):
        m.unmount(Gio.MountUnmountFlags.NONE, None, self._eject_cb, None)

    def load_preferences(self):
        configuration = Configuration()
        self.show_hdd = configuration.get('show-hdd')
        self.show_net = configuration.get('show-net')

    def save_config(self):
        configuration = Configuration()
        configuration.set('show-hdd', self.show_hdd)
        configuration.set('show-net', self.show_net)
        configuration.save()
        self.refresh()

    def update(self):
        if len(self.drives) > 0:
            self.indicator.set_status(AppIndicator.IndicatorStatus.ACTIVE)
        else:
            self.indicator.set_status(AppIndicator.IndicatorStatus.PASSIVE)
        if self.menu is not None:
            self.menu.destroy()
        self.menu = Gtk.Menu()
        # Header
        hdr = Gtk.MenuItem(label=_('Safely remove devices')+':')
        hdr.show()
        self.menu.append(hdr)
        # Separator
        smi = Gtk.SeparatorMenuItem()
        smi.show()
        self.menu.append(smi)
        for i in self.drives:
            mi = Gtk.ImageMenuItem(label=i.get_name())
            # XXX: gicon does not work...
            mi.set_image(Gtk.Image.new_from_stock(
                _get_icon_name_from_gicon(i.get_icon()), Gtk.IconSize.MENU))
            mi.set_always_show_image(True)
            self.menu.append(mi)
            mi.show()
            mi.connect('activate', self.eject, i)
        separator = Gtk.SeparatorMenuItem()
        separator.show()
        self.menu.append(separator)
        preferences = Gtk.MenuItem(label=_('Preferences'))
        preferences.connect('activate', self.on_preferences_activated)
        preferences.show()
        self.menu.append(preferences)
        separator = Gtk.SeparatorMenuItem()
        separator.show()
        self.menu.append(separator)
        about_item = Gtk.MenuItem(label=_('About...'))
        about_item.show()
        about_item.connect('activate', self.on_about_activate)
        self.menu.append(about_item)
        separator = Gtk.SeparatorMenuItem()
        separator.show()
        self.menu.append(separator)
        quit_item = Gtk.MenuItem(label=_('Exit'))
        quit_item.show()
        quit_item.connect('activate', self.on_exit_activate)
        self.menu.append(quit_item)
        self.indicator.set_menu(self.menu)

    def on_exit_activate(self, widget):
        exit(0)

    def on_preferences_activated(self, widget):
        cm = PreferencesDialog(None)
        if cm.run() == Gtk.ResponseType.ACCEPT:
            cm.save_preferences()
            self.update()
        cm.hide()
        cm.destroy()

    def on_about_activate(self, widget):
        widget.set_sensitive(False)
        ad = Gtk.AboutDialog()
        ad.set_name(comun.APPNAME)
        ad.set_version(comun.VERSION)
        ad.set_copyright('''
Copyrignt (c) 2010-2016

Pedro Fragoso <ember@pfragoso.org>
Calum Lind <https://github.com/cas-->
Anton Eliasson <devel@antoneliasson.se>
Lorenzo Carbonell Cerezo <lorenzo.carbonell.cerezo@gmail.com>
''')
        ad.set_comments(_('Cool flash drive unmounter for Ubuntu'))
        ad.set_license('''
This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
''')
        ad.set_website('http://www.atareao.es')
        ad.set_website_label('http://www.atareao.es')
        ad.set_authors([
            'Pedro Fragoso <ember@pfragoso.org>',
            'Calum Lind <https://github.com/cas-->',
            'Anton Eliasson <devel@antoneliasson.se>',
            'Lorenzo Carbonell Cerezo <lorenzo.carbonell.cerezo@gmail.com>'])
        ad.set_translator_credits('''
Lorenzo Carbonell Cerezo <lorenzo.carbonell.cerezo@gmail.com>\n\
''')
        ad.set_documenters([
            'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
        ad.set_logo(GdkPixbuf.Pixbuf.new_from_file(comun.IMAGE))
        ad.set_icon(GdkPixbuf.Pixbuf.new_from_file(comun.ICON))
        ad.set_program_name(comun.APPNAME)
        ad.run()
        ad.destroy()
        widget.set_sensitive(True)

    def on_option(self, x, o):
        if o == 'hdd':
            self.monitor.show_hdd = x.get_active()
        if o == 'net':
            self.monitor.show_net = x.get_active()
        self.save_config()


def main():
    Notify.init('indicator-usb')
    m = Main()
    Gtk.main()

if __name__ == '__main__':
    main()
